[cxx-interop]: Enhance CxxSet with query and set operations

CrazyFanFan opened the pull request:
This commit enriches the `CxxSet` protocol with additional query and set operations, including union, intersection, subtraction, and subset/superset checks. These enhancements align the C++ set wrapper with Swift's standard library set operations, providing a more comprehensive and versatile API.

I am not entirely sure if these additional operations should be included or if there are plans to support calling std::algorithm directly from Swift in the future. However, since the `remove` operation is already supported, I decided to try and add a few more.

egorzhdan commented:
@swift-ci please smoke test

egorzhdan reviewed (commented):
Awesome, thanks a lot @CrazyFanFan!
I think it would be great to have this working on Linux before we merge the PR, but the overall change looks good to me. Thanks for doing the legwork here.

egorzhdan commented:
@swift-ci please test

CrazyFanFan commented:
@egorzhdan Can this PR be merged, or is there anything else that needs to be done?

egorzhdan commented:
@swift-ci please test

CrazyFanFan commented on stdlib/public/Cxx/CxxSet.swift:
```diff
@@ -74,17 +77,43 @@ public protocol CxxUniqueSet<Element>: CxxSet {
+  func __endUnsafe() -> RawIterator
+  func __findUnsafe(_ value: Element) -> RawIterator
+
+  #if !os(Linux)
```
Thank you for your comment!

The API behavior of C++ `std::set `and `std::unordered_set` is inconsistent between Linux and macOS. I couldn't find a unified API implementation, so I removed support for Linux through conditional compilation.

I'll try to resolve this discrepancy.

CrazyFanFan replied:
@egorzhdan  The methods below are not implemented in `std::set` and `std::unordered_set` on Linux. I think I might have missed some details in the ClangImporter. I haven't found a solution to make this code work well on Linux for now. Do you have any suggestions?
```swift
Copy
mutating func __endMutatingUnsafe() -> RawMutableIterator
mutating func __findMutatingUnsafe(_ value: Element) -> RawMutableIterator

@discardableResult
mutating func __eraseUnsafe(_ iter: RawMutableIterator) -> RawMutableIterator
```

CrazyFanFan committed:
[cxx-interop]: Enhance CxxSet with query and set operations

This commit enriches the `CxxSet` protocol with additional query and set operations, including union, intersection, subtraction, and subset/superset checks. These enhancements align the C++ set wrapper with Swift's standard library set operations, providing a more comprehensive and versatile API.

CrazyFanFan commented:
This PR is now updated and ready for review. I've removed the Linux-specific conditional compilation and verified all tests pass locally on both Linux and macOS.
